enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,370 @@
+feature in that context.  We aim to fill this gap with an optional range of
+capabilities.
+
+### Goals
+
+A SLB solution must provide these high level features:
+
+* Management of one or more pools of IP addresses to be allocated for SLBs
+* High Availability (HA) management of these IP addresses once allocated.  We
+  must be able to fail over addresses in less than 5 seconds.
+* Advertisement of the location of an IP address once assigned to a Node or
+  moved to a new Node.  Possible approaches include:
+    * Layer 2 (gratuitous ARP for IPv4, NDP for IPv6) - good for wide range of
+      environment compatibility, but limiting for larger clusters
+    * Layer 3 (BGP) - good for integration with networks for larger clusters
+* Suitable for large scale clusters (target up to 2000 nodes).
+* Single Node IPs (all traffic for a Service LB going through a single Node) is
+  acceptable for our out-of-the-box solution.  Support for a multi-Node based
+  solution would be nice, but is not required.
+* Must be compatible with at least the following cluster network types:
+  [OpenShift-SDN](https://github.com/openshift/sdn) and
+  [OVN-Kubernetes](https://github.com/openshift/ovn-kubernetes)
+
+### Non-Goals
+
+* We can also support this functionality through the use of partner add-ons,
+  but discussion of those solutions is out of scope for this document.
+
+## Proposal
+
+Adopt [MetalLB](https://metallb.universe.tf/) as an out-of-the-box solution for
+most on-premise SLB use cases.
+
+MetalLB is commonly referenced when people discuss service load balancers for
+bare metal.  The [concepts page](https://metallb.universe.tf/concepts/) gives a
+pretty good overview of how it works.  It manages pools of IP addresses to
+allocate for SLBs.  Once an IP is allocated to a SLB, it’s assigned to a Node
+and the location of that IP address must be announced externally.  It has two
+modes to announce IPs: [layer 2](https://metallb.universe.tf/concepts/layer2/)
+(ARP for IPv4, NDP for IPv6) or
+[BGP](https://metallb.universe.tf/concepts/bgp/).  The layer 2 mode is
+sufficient for smaller scale clusters, while the BGP mode can work at much
+larger scale.
+
+While the BGP option is attractive for scaling reasons, it’s also more
+complicated and will not work in all environments.  It won’t work in an
+environment that won’t allow BGP advertisements from the cluster Nodes.  If a
+cluster uses a Network addon that also makes use of BGP, MetalLB integration
+will be more challenging.  For example see the [MetalLB page about Calico
+support](https://metallb.universe.tf/configuration/calico/).
+
+The layer 2 mode has the advantage of working in more environments.  We could
+also consider a MetalLB enhancement that makes it understand different L2
+domains and manage different IP address pools for each domain where SLBs may
+reside.
+
+### User Stories
+
+#### Story 1 - Easy Use with a Small Cluster
```

Comment:
We should define maybe Story 0 that is "As an application deployer on top of metal, a bare metal UPI cluster supports service load balancers exactly like a cloud service load balancer does, including in cluster access, out of cluster access, the possibility of DNS, the e2e tests working exactly like they do for AWS, etc etc etc".  I think it's important to define that the goal of this is to at minimum be consistent with cloud, then broaden into domain specific problems.

Reply:
We can't make _all_ bare metal UPI clusters work like cloud environments (eg, there's no plan for clusters where the nodes are on multiple L2 segments and you can't use BGP between them). But Russell could merge the "exactly like a cloud service" text into his existing stories.

Reply:
I'll look at how to incorporate the "exactly like a cloud service" input into the text.

Regarding clusters on multiple L2 segments and no BGP available, I think that's an enhancement someone could implement.  I've seen it requested on issues, and someone wrote a PR to do it (https://github.com/metallb/metallb/pull/502), though it needs some further work.  I'd like to see something done for this, because it seems like a common enough use case.

Reply:
I've captured this, thanks